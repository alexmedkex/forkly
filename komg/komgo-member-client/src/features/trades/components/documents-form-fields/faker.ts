import { mockCategories } from '../../../document-management/store/categories/mock-data'
import { mockProduct } from '../../../document-management/store/products/mock-data'
import { DocumentType } from '../../../document-management/store/types/document-type'

const fakeCommercialContractDocumentType: DocumentType = {
  product: mockProduct,
  category: mockCategories[0],
  name: 'Commercial Contract',
  id: 'commercialContract',
  fields: [],
  predefined: false
}

export const fakeCommercialContractDocumentTypes: DocumentType[] = [fakeCommercialContractDocumentType]

export const fakeCommercialContractDocumentWithoutExtension = {
  _id: 'document0',
  file: {
    size: 5000
  },
  name: 'documentName',
  fileName: 'documentFileName',
  typeId: 'commercialContract',
  fileType: 'application/something',
  type: {
    id: 'commercialContract'
  }
}

export const fakeCommercialContractDocumentWithExtension = {
  _id: 'document1',
  file: {
    size: 5000
  },
  name: 'documentName1',
  fileName: 'documentFileName1.pdf',
  typeId: 'commercialContract',
  fileType: 'application/pdf',
  type: {
    id: 'commercialContract'
  }
}
