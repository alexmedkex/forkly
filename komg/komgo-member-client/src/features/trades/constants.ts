import { ITradeDocument } from './store/types'
import {
  TradeSource,
  PriceUnit,
  PaymentTermsTimeUnit,
  PaymentTermsDayType,
  PaymentTermsWhen,
  PaymentTermsEventBase,
  Currency,
  ICargoBase,
  ITradeBase,
  CreditRequirements,
  CARGO_SCHEMA_VERSION,
  PARCEL_SCHEMA_VERSION,
  TRADE_SCHEMA_VERSION
} from '@komgo/types'
import { IParcelWithOverrides } from './components/cargo-form-fields/ParcelData'
import { v4 } from 'uuid'

export const NOTENOUGHINFO = 'Not enough information provided.'
export const UNRECOGNISED_STATUS = 'Status unrecognized'

export const LOC_STATUS = {
  TO_BE_FINANCED: 'TO_BE_FINANCED',
  SUBMITTED: 'SUBMITTED',
  RECEIVED: 'RECEIVED',
  ADVISED: 'ADVISED',
  AMENDMENT_REQUESTED: 'AMENDMENT_REQUESTED',
  AMENDMENT_UNDER_NEGOTIATION: 'AMENDMENT_UNDER_NEGOTIATION',
  AMENDMENT_ACCEPTED: 'AMENDMENT_ACCEPTED',
  PAID: 'PAID',
  REFUSED: 'REFUSED',
  EXPIRED: 'EXPIRED'
}

export enum LOC_CATEGORIES {
  TO_BE_FINANCED = 'TO_BE_FINANCED',
  IN_PROGRESS = 'IN_PROGRESS',
  ISSUED = 'ISSUED',
  CLOSED = 'CLOSED'
}

export enum TRADE_STATUS {
  ToBeFinanced = 'TO_BE_FINANCED',
  ToBeDiscounted = 'TO_BE_DISCOUNTED'
}

export const LOC_CATEGORY_COLOURS = {
  [LOC_CATEGORIES.TO_BE_FINANCED]: 'red',
  [LOC_CATEGORIES.IN_PROGRESS]: 'yellow',
  [LOC_CATEGORIES.ISSUED]: 'green'
}

export const READABLE_LOC_STATUS = {
  [LOC_CATEGORIES.TO_BE_FINANCED]: [LOC_STATUS.TO_BE_FINANCED],
  [LOC_CATEGORIES.IN_PROGRESS]: [LOC_STATUS.SUBMITTED],
  [LOC_CATEGORIES.ISSUED]: [
    LOC_STATUS.RECEIVED,
    LOC_STATUS.ADVISED,
    LOC_STATUS.AMENDMENT_ACCEPTED,
    LOC_STATUS.AMENDMENT_UNDER_NEGOTIATION,
    LOC_STATUS.AMENDMENT_ACCEPTED
  ],
  [LOC_CATEGORIES.CLOSED]: [LOC_STATUS.PAID, LOC_STATUS.REFUSED, LOC_STATUS.EXPIRED]
}

export const DESC = -1
export const ASC = 1
export const DESC_TEXT = 'descending'
export const ASC_TEXT = 'ascending'

export enum TradingRole {
  SELLER = 'seller',
  BUYER = 'buyer'
}

export const TRADING_ROLE_OPTIONS = {
  BUYER: TradingRole.BUYER.toUpperCase(),
  SELLER: TradingRole.SELLER.toUpperCase()
}

export const CREDIT_REQUIREMENT_LABEL = {
  DOCUMENTARY_LETTER_OF_CREDIT: 'Documentary LC',
  STANDBY_LETTER_OF_CREDIT: 'Standby LC',
  OPEN_CREDIT: 'Open Credit',
  OFFSET: 'Offset'
}

export const CREDIT_REQUIREMENT_DISPLAY_VALUES = {
  [CreditRequirements.DocumentaryLetterOfCredit]: [CREDIT_REQUIREMENT_LABEL.DOCUMENTARY_LETTER_OF_CREDIT],
  [CreditRequirements.StandbyLetterOfCredit]: [CREDIT_REQUIREMENT_LABEL.STANDBY_LETTER_OF_CREDIT],
  [CreditRequirements.OpenCredit]: [CREDIT_REQUIREMENT_LABEL.OPEN_CREDIT],
  [CreditRequirements.Offset]: [CREDIT_REQUIREMENT_LABEL.OFFSET]
}

export const tradeDefaultEmptyValues: Partial<ITradeBase> = {
  buyerEtrmId: '',
  sellerEtrmId: '',
  dealDate: '',
  seller: '',
  buyer: '',
  commodity: '',
  generalTermsAndConditions: '',
  deliveryPeriod: { startDate: '', endDate: '' },
  demurrageTerms: '',
  laytime: '',
  law: '',
  creditRequirement: '' as any,
  contractReference: '',
  priceFormula: ''
}

export const initialTradeData: ITradeBase = {
  version: TRADE_SCHEMA_VERSION.V2,
  ...(tradeDefaultEmptyValues as ITradeBase),
  source: TradeSource.Komgo,
  dealDate: '',
  priceOption: undefined,
  price: undefined,
  currency: Currency.USD,
  priceUnit: PriceUnit.BBL,
  quantity: undefined,
  minTolerance: undefined,
  maxTolerance: undefined,
  invoiceQuantity: undefined,
  paymentTermsOption: undefined,
  paymentTerms: undefined,
  deliveryTerms: undefined,
  deliveryLocation: undefined,
  requiredDocuments: []
}

export const initialCargoData: ICargoBase = {
  source: TradeSource.Komgo,
  grade: '',
  cargoId: '',
  parcels: [],
  sourceId: '',
  originOfGoods: '',
  quality: '',
  version: CARGO_SCHEMA_VERSION.V2
}

export const generateInitialParcelData = (): IParcelWithOverrides => {
  const uniqueKey = v4()

  // As we have all parcels stored in a list in formik, when a new parcel is created
  // we need a unique key to track which item in the list is being edited
  return { ...initialParcelData, uniqueKey }
}

export const initialParcelData: IParcelWithOverrides = {
  id: '',
  laycanPeriod: {
    startDate: '',
    endDate: ''
  },
  modeOfTransport: '',
  modeOfTransportOther: '',
  vesselIMO: 0,
  vesselName: '',
  loadingPort: '',
  dischargeArea: '',
  inspector: '',
  deemedBLDate: '',
  quantity: undefined,
  warehouseOperatorName: '',
  pipelineName: '',
  tankFarmOperatorName: '',
  destinationPlace: '',
  loadingPlace: '',
  version: PARCEL_SCHEMA_VERSION.V2
}

export const initialDocumentData: ITradeDocument = {
  name: '',
  typeId: '',
  file: null,
  fileName: '',
  fileType: ''
}

export const TRADE_EDITED_MESSAGE = 'Trade edited'
export const TRADE_CREATED_MESSAGE = 'New trade created'
export const TRADE_AND_CARGO_EDITED_MESSAGE = 'Trade and cargo edited'
export const TRADE_DELETED_MESSAGE = 'Trade deleted'

export const emptyDropdownItem = { text: '(none)', content: '(none)' }
