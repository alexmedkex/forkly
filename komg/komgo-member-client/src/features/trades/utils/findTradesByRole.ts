import { ITradeEnriched } from '../store/types'
import { TradingRole } from '../constants'

export const findTradesByRole = (trades: ITradeEnriched[], tradingRole: TradingRole, company: string) =>
  trades.filter(trade => trade[tradingRole] === company)
