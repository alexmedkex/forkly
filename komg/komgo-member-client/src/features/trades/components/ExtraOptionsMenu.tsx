import * as React from 'react'
import { ITrade } from '@komgo/types'

import { WithPermissionsProps, withPermissions } from '../../../components'
import { connect } from 'react-redux'
import { compose } from 'redux'
import { tradeFinanceManager, Permission } from '@komgo/permissions'
import { productRD } from '@komgo/products'
import { Dimmer, Loader, Dropdown } from 'semantic-ui-react'
import { RouteComponentProps, withRouter } from 'react-router-dom'
import { ApplicationState } from '../../../store/reducers'
import { ILetterOfCredit } from '../../letter-of-credit-legacy/types/ILetterOfCredit'
import { WithLicenseCheckProps, withLicenseCheck } from '../../../components/with-license-check'
import { loadingSelector } from '../../../store/common/selectors'
import { LetterOfCreditActionType as LegacyLetterOfCreditActionType } from '../../letter-of-credit-legacy/store/types'
import { IStandbyLetterOfCredit, IReceivablesDiscountingInfo } from '@komgo/types'
import { TradingRole } from '../constants'
import { getLatestFinancialInstrumentsForTrade } from '../utils/selectors'
import { getTradeActionsForFinancialInstruments, TradeAction, canEditTrade } from '../utils/tradeActionUtils'
import {
  ExtraOptionsMenuForRDProps,
  createRDApplyForDiscountingLink,
  isRDApplyForDiscountingValid
} from '../../receivable-finance/pages/dashboard/components/ExtraOptionsMenu'
import { StandbyLetterOfCreditActionType } from '../../standby-letter-of-credit-legacy/store/types'
import { ReceivableDiscountingApplicationActionType } from '../../receivable-discounting-legacy/store/application/types'
import { Strings } from '../../receivable-discounting-legacy/resources/strings'
import { stringify } from 'qs'
import { ILetterOfCreditWithData, LetterOfCreditActionType } from '../../letter-of-credit/store/types'
import { ITradeEnriched } from '../store/types'

export interface ExtraOptionsMenuOwnProps {
  tradeId: string
  sourceId: string
  role: TradingRole
}

export interface ExtraOptionsMenuAllProps
  extends WithPermissionsProps,
    ExtraOptionsMenuOwnProps,
    WithLicenseCheckProps,
    RouteComponentProps<any> {
  trade: ITradeEnriched
  letterOfCredit?: ILetterOfCredit
  standbyLetterOfCredit?: IStandbyLetterOfCredit
  newLetterOfCredit?: ILetterOfCreditWithData
  rdMenuProps?: ExtraOptionsMenuForRDProps
  isFetching: boolean
}

export const DISABLED: any = {
  color: 'currentColor',
  cursor: 'not-allowed',
  pointerEvents: 'none',
  opacity: 0.5,
  textDecoration: 'none'
}

interface IOption {
  label: string
  name: string
  buildLink: (props: ExtraOptionsMenuAllProps) => string
  permission: Permission
  isValid: (props: ExtraOptionsMenuAllProps) => boolean
}

const VIEW_OPTION: IOption = {
  label: 'View Trade Details',
  name: 'details',
  buildLink: (props: ExtraOptionsMenuAllProps) => `/trades/${props.trade!._id}`,
  permission: tradeFinanceManager.canReadTrades,
  isValid: () => true
}
const EDIT_OPTION: IOption = {
  label: 'Edit Trade',
  name: 'editTrade',
  buildLink: (props: ExtraOptionsMenuAllProps) => `/trades/${props.trade._id}/edit`,
  permission: tradeFinanceManager.canCrudTrades,
  isValid: (props: ExtraOptionsMenuAllProps) =>
    canEditTrade(props.trade, props.rdMenuProps && props.rdMenuProps.rdStatus)
}

const TRADE_OPTIONS = [VIEW_OPTION, EDIT_OPTION]

const getCreateLCRoute = (trade: ITrade) => {
  return `/financial-instruments/letters-of-credit/new?tradeId=${trade!._id}`
}

const getCreateSBLCRoute = (trade: ITrade) => {
  return `/financial-instruments/standby-letters-of-credit/new?source=${trade!.source}&sourceId=${trade!.sourceId}`
}

// TEMP HACK

export const getCreateNewSBLCRoute = (trade: ITrade) => {
  const params = stringify({
    select: true,
    redirectTo: `/letters-of-credit/new?source=${trade!.source}&sourceId=${trade!.sourceId}`,
    type: 'SBLC'
  })
  return `/templates?${params}`
}

// END HACK

const isFinancialInstrumentActionAllowed = (props: ExtraOptionsMenuAllProps, action: TradeAction) => {
  return getTradeActionsForFinancialInstruments(
    props.trade,
    props.role,
    props.letterOfCredit,
    props.standbyLetterOfCredit,
    props.newLetterOfCredit
  ).includes(action)
}

const FINANCIAL_INSTRUMENT_OPTIONS: IOption[] = [
  {
    label: 'Apply for LC',
    name: 'applyforLc',
    buildLink: (props: ExtraOptionsMenuAllProps) => getCreateLCRoute(props.trade),
    permission: tradeFinanceManager.canManageLCRequests,
    isValid: (props: ExtraOptionsMenuAllProps) => isFinancialInstrumentActionAllowed(props, TradeAction.ApplyForLC)
  },
  {
    label: 'Reapply for LC',
    name: 'reapplyforLc',
    buildLink: (props: ExtraOptionsMenuAllProps) => getCreateLCRoute(props.trade),
    permission: tradeFinanceManager.canManageLCRequests,
    isValid: (props: ExtraOptionsMenuAllProps) => isFinancialInstrumentActionAllowed(props, TradeAction.ReapplyForLC)
  },
  {
    label: 'View LC Details',
    name: 'viewLc',
    buildLink: (props: ExtraOptionsMenuAllProps) =>
      props.letterOfCredit ? `/financial-instruments/letters-of-credit/${props.letterOfCredit._id}` : '#',
    permission: tradeFinanceManager.canManageLCRequests,
    isValid: (props: ExtraOptionsMenuAllProps) => isFinancialInstrumentActionAllowed(props, TradeAction.ViewLC)
  },
  {
    label: 'View Legacy SBLC Details',
    name: 'viewSBLC',
    buildLink: (props: ExtraOptionsMenuAllProps) =>
      props.standbyLetterOfCredit
        ? `/financial-instruments/standby-letters-of-credit/${props.standbyLetterOfCredit.staticId}`
        : '#',
    permission: tradeFinanceManager.canManageLCRequests,
    isValid: (props: ExtraOptionsMenuAllProps) => isFinancialInstrumentActionAllowed(props, TradeAction.ViewLegacySBLC)
  },
  {
    label: 'Apply for SBLC',
    name: 'applyfornewSBLC',
    buildLink: (props: ExtraOptionsMenuAllProps) => getCreateNewSBLCRoute(props.trade),
    permission: tradeFinanceManager.canManageLCRequests,
    isValid: (props: ExtraOptionsMenuAllProps) => isFinancialInstrumentActionAllowed(props, TradeAction.ApplyForSBLC)
  },
  {
    label: 'Reapply for SBLC',
    name: 'reapplyfornewSBLC',
    buildLink: (props: ExtraOptionsMenuAllProps) => getCreateNewSBLCRoute(props.trade),
    permission: tradeFinanceManager.canManageLCRequests,
    isValid: (props: ExtraOptionsMenuAllProps) => isFinancialInstrumentActionAllowed(props, TradeAction.ReapplyForSBLC)
  },
  {
    label: 'View SBLC Details',
    name: 'viewNewSBLC',
    buildLink: (props: ExtraOptionsMenuAllProps) =>
      props.newLetterOfCredit ? `/letters-of-credit/${props.newLetterOfCredit.staticId}` : '#',
    permission: tradeFinanceManager.canManageLCRequests,
    isValid: (props: ExtraOptionsMenuAllProps) => isFinancialInstrumentActionAllowed(props, TradeAction.ViewSBLC)
  }
]

const BUYER_OPTIONS = [...TRADE_OPTIONS, ...FINANCIAL_INSTRUMENT_OPTIONS]

const RD_APPLY_FOR_DISCOUNTING_OPTION: IOption = {
  label: Strings.ApplyForDiscountingRiskCoverActionText,
  name: 'applyforDiscounting',
  permission: tradeFinanceManager.canReadRD,
  buildLink: (props: ExtraOptionsMenuAllProps) => createRDApplyForDiscountingLink(props.tradeId),
  isValid: (props: ExtraOptionsMenuAllProps) =>
    isSeller(props.role) && isRDApplyForDiscountingValid(props.rdMenuProps) && props.isLicenseEnabled(productRD)
}

const TRADE_DASHBOARD_RD_OPTIONS: IOption[] = [VIEW_OPTION, EDIT_OPTION, RD_APPLY_FOR_DISCOUNTING_OPTION]

const isSeller = (role: TradingRole) => role === TradingRole.SELLER
const isBuyer = (role: TradingRole) => role === TradingRole.BUYER

export const ExtraOptionsMenuFactory = (props: ExtraOptionsMenuAllProps): JSX.Element[] => {
  const { isFetching, isAuthorized, role, history } = props
  const options = role === TradingRole.BUYER ? BUYER_OPTIONS.filter(o => o.isValid(props)) : TRADE_DASHBOARD_RD_OPTIONS

  const usedOptions = options.map((option: IOption) => ({
    ...option,
    disabled: !isAuthorized(option.permission) || !option.isValid(props)
  }))
  const availableOptions = usedOptions.map(({ disabled, name, buildLink, label }, idx) => (
    <Dropdown.Item
      key={`${name}-${idx}`}
      data-test-id={name}
      disabled={disabled}
      name={name}
      onClick={() => history.push(buildLink(props))}
    >
      {label}
    </Dropdown.Item>
  ))

  return isFetching
    ? [
        <Dimmer active={true} inverted={true} key={'dimmer'}>
          <Loader inverted={true} size="mini" />
        </Dimmer>
      ]
    : availableOptions
}

const mapStateToProps = (
  state: ApplicationState,
  ownProps: ExtraOptionsMenuOwnProps
): Partial<ExtraOptionsMenuAllProps> => {
  let letterOfCredit: ILetterOfCredit
  let standbyLetterOfCredit: IStandbyLetterOfCredit
  let newLetterOfCredit: ILetterOfCreditWithData
  let rdMenuProps: ExtraOptionsMenuForRDProps
  let isFetching

  const trade = state
    .get('trades')
    .get('trades')
    .toJS()[ownProps.tradeId]

  if (ownProps.role === TradingRole.BUYER) {
    ;({ letterOfCredit, standbyLetterOfCredit, newLetterOfCredit } = getLatestFinancialInstrumentsForTrade(
      state,
      trade
    ))
    isFetching = loadingSelector(state.get('loader').get('requests'), [
      LegacyLetterOfCreditActionType.LETTERS_OF_CREDIT_REQUEST,
      StandbyLetterOfCreditActionType.FETCH_STANDBY_LETTERS_OF_CREDIT_REQUEST,
      LetterOfCreditActionType.FETCH_LETTERS_OF_CREDIT_REQUEST
    ])
  } else if (ownProps.role === TradingRole.SELLER) {
    const rdInfo: IReceivablesDiscountingInfo = state
      .get('receivableDiscountingApplication')
      .get('byId')
      .toList()
      .toJS()
      .find((rdInfo: IReceivablesDiscountingInfo) => rdInfo.rd.tradeReference.sourceId === ownProps.sourceId)
    if (rdInfo) {
      rdMenuProps = {
        rdId: rdInfo.rd.staticId,
        rdStatus: rdInfo.status
      }
    }
    isFetching = loadingSelector(state.get('loader').get('requests'), [
      ReceivableDiscountingApplicationActionType.FETCH_MULTIPLE_APPLICATION_REQUEST
    ])
  }

  return {
    letterOfCredit,
    standbyLetterOfCredit,
    newLetterOfCredit,
    trade,
    rdMenuProps,
    isFetching
  }
}

export default compose<any>(withPermissions, withLicenseCheck, withRouter, connect(mapStateToProps))(
  ExtraOptionsMenuFactory
)
