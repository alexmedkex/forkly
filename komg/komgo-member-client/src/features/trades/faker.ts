import { Document, Product } from '../document-management'
import { Products } from '../document-management/constants/Products'
import { fakeDocument } from '../document-management/utils/faker'

export const fakeUploadedDocument = ({ id = '1', name = 'documentName', typeName = 'commercialContract' } = {}) => {
  const mockProduct: Product = { name: 'TRADE FINANCE', id: Products.TradeFinance }
  const mockCategory = { id: 'commercial-contract', name: 'commercial-contract', product: mockProduct }
  const mockType = {
    id: '2',
    name: typeName,
    product: mockProduct,
    category: mockCategory,
    fields: [],
    predefined: true
  }
  const COMPANY_ID = 'societegenerale'
  const doc: Document = fakeDocument({
    id,
    documentId: 'document-id',
    name,
    product: mockProduct,
    category: mockCategory,
    type: mockType,
    owner: { firstName: 'Owner', lastName: 'Owner', companyId: 'company-id' },
    hash: 'hash',
    registrationDate: new Date('2019-02-04T14:47:26.988Z'),
    metadata: [],
    content: undefined,
    sharedWith: [{ counterpartyId: COMPANY_ID, sharedDates: [new Date('2019-02-04T14:47:26.988Z')] }],
    sharedBy: ''
  })
  return doc
}
