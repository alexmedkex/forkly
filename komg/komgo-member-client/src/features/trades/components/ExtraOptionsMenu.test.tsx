import * as React from 'react'
import { ExtraOptionsMenuFactory, ExtraOptionsMenuAllProps } from './ExtraOptionsMenu'
import { fakeLetterOfCredit, fakeTrade } from '../../letter-of-credit-legacy/utils/faker'
import { BrowserRouter as Router, Link } from 'react-router-dom'
import * as renderer from 'react-test-renderer'
import { TradingRole, TRADE_STATUS } from '../constants'
import { ShallowWrapper } from 'enzyme'
import { RDStatus, TradeSource, CreditRequirements } from '@komgo/types'
import { createMemoryHistory } from 'history'
import { Dropdown } from 'semantic-ui-react'
import { ILetterOfCreditStatus } from '../../letter-of-credit-legacy/types/ILetterOfCredit'

const company = 'myCompany'
const testProps: ExtraOptionsMenuAllProps = {
  isFetching: false,
  isLicenseEnabledForCompany: () => true,
  letterOfCredit: fakeLetterOfCredit(),
  trade: {
    ...fakeTrade(),
    sourceId: '1',
    _id: 'abc',
    source: TradeSource.Vakt,
    buyer: company
  },
  tradeId: 'abc',
  sourceId: '1',
  role: TradingRole.BUYER,
  isAuthorized: () => true,
  isLicenseEnabled: () => true,
  history: createMemoryHistory(),
  location: {
    pathname: '',
    search: '',
    state: '',
    hash: ''
  },
  match: {
    isExact: true,
    path: '',
    url: '',
    params: null
  },
  staticContext: undefined
}

describe('ExtraOptionsMenu', () => {
  describe('as buyer or seller', () => {
    it('shows view Trade', () => {
      const tree = renderer
        .create(
          <Router>
            <Dropdown.Menu>{ExtraOptionsMenuFactory({ ...testProps, letterOfCredit: undefined })}</Dropdown.Menu>
          </Router>
        )
        .toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  describe('as buyer', () => {
    describe('ApplyLC', () => {
      it('shows', () => {
        const tree = renderer
          .create(
            <Router>
              <Dropdown.Menu>{ExtraOptionsMenuFactory({ ...testProps, letterOfCredit: undefined })}</Dropdown.Menu>
            </Router>
          )
          .toJSON()
        expect(tree).toMatchSnapshot()
      })

      it('is disabled', () => {
        const tree = renderer
          .create(
            <Router>
              <Dropdown.Menu>{ExtraOptionsMenuFactory({ ...testProps })}</Dropdown.Menu>
            </Router>
          )
          .toJSON()
        expect(tree).toMatchSnapshot()
      })
    })

    describe('Reapply for LC', () => {
      it(`shows if status is ${ILetterOfCreditStatus.REQUEST_REJECTED}`, () => {
        const tree = renderer
          .create(
            <Router>
              <Dropdown.Menu>
                {ExtraOptionsMenuFactory({
                  ...testProps,
                  letterOfCredit: { ...fakeLetterOfCredit(), status: ILetterOfCreditStatus.REQUEST_REJECTED }
                })}
              </Dropdown.Menu>
            </Router>
          )
          .toJSON()
        expect(tree).toMatchSnapshot()
      })

      it(`shows if status is ${ILetterOfCreditStatus.ISSUED_LC_REJECTED}`, () => {
        const tree = renderer
          .create(
            <Router>
              <Dropdown.Menu>
                {ExtraOptionsMenuFactory({
                  ...testProps,
                  letterOfCredit: { ...fakeLetterOfCredit(), status: ILetterOfCreditStatus.ISSUED_LC_REJECTED }
                })}
              </Dropdown.Menu>
            </Router>
          )
          .toJSON()
        expect(tree).toMatchSnapshot()
      })

      it('is disabled', () => {
        const tree = renderer
          .create(
            <Router>
              <Dropdown.Menu>{ExtraOptionsMenuFactory({ ...testProps, letterOfCredit: undefined })}</Dropdown.Menu>
            </Router>
          )
          .toJSON()
        expect(tree).toMatchSnapshot()
      })
    })

    describe('View LC Details', () => {
      it('shows', () => {
        const tree = renderer
          .create(
            <Router>
              <Dropdown.Menu>{ExtraOptionsMenuFactory({ ...testProps })}</Dropdown.Menu>
            </Router>
          )
          .toJSON()
        expect(tree).toMatchSnapshot()
      })

      it('is disabled', () => {
        const tree = renderer
          .create(
            <Router>
              <Dropdown.Menu>{ExtraOptionsMenuFactory({ ...testProps, letterOfCredit: undefined })}</Dropdown.Menu>
            </Router>
          )
          .toJSON()
        expect(tree).toMatchSnapshot()
      })
    })
  })

  describe('as seller', () => {
    describe('Apply for Discounting Dashboard options', () => {
      it('shows view and edit trade, and apply for discounting buttons', () => {
        const push = jest.fn()
        const wrapper = new ShallowWrapper(
          (
            <Dropdown.Menu>
              {ExtraOptionsMenuFactory({
                ...testProps,
                role: TradingRole.SELLER,
                history: { ...testProps.history, push }
              })}
            </Dropdown.Menu>
          )
        )

        const apply = wrapper.find("[data-test-id='applyforDiscounting']")
        const view = wrapper.find("[data-test-id='details']")
        const edit = wrapper.find("[data-test-id='editTrade']")

        expect(apply.exists()).toBeTruthy()
        expect(view.exists()).toBeTruthy()
        expect(edit.exists()).toBeTruthy()
        expect(edit.props().disabled).toBeTruthy()
        expect(wrapper.find(Dropdown.Item).length).toBe(3)

        apply.simulate('click')
        expect(push).toHaveBeenCalledWith('/receivable-discounting/abc/apply')
      })

      it('should allow edit for seller if status is TO_BE_DISCOUNTED', () => {
        const push = jest.fn()

        const props: ExtraOptionsMenuAllProps = {
          ...testProps,
          trade: {
            ...testProps.trade,
            source: TradeSource.Komgo,
            creditRequirement: CreditRequirements.OpenCredit,
            status: TRADE_STATUS.ToBeDiscounted
          },
          history: {
            ...testProps.history,
            push
          }
        }

        const wrapper = new ShallowWrapper(
          <Dropdown.Menu>{ExtraOptionsMenuFactory({ ...props, role: TradingRole.SELLER })}</Dropdown.Menu>
        )

        const apply = wrapper.find({ name: 'editTrade' })
        expect(apply.props().disabled).toBeFalsy()
        apply.simulate('click')

        expect(push).toHaveBeenCalledWith(`/trades/${props.trade._id}/edit`)
      })

      it('should not allow edit for seller if status is not TO_BE_DISCOUNTED', () => {
        const props = {
          ...testProps,
          trade: {
            ...testProps.trade,
            source: TradeSource.Komgo,
            creditRequirement: CreditRequirements.OpenCredit,
            tradingRole: TradingRole.SELLER,
            status: 'REQUESTED'
          }
        }

        const wrapper = new ShallowWrapper(
          <Dropdown.Menu>{ExtraOptionsMenuFactory({ ...props, role: TradingRole.SELLER })}</Dropdown.Menu>
        )

        const apply = wrapper.find({ name: 'editTrade' })
        expect(apply.props().disabled).toBeTruthy()
      })
    })
  })
})
