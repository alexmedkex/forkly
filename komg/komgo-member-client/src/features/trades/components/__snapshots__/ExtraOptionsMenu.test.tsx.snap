// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExtraOptionsMenu as buyer ApplyLC is disabled 1`] = `
<div
  className="menu transition"
>
  <div
    aria-disabled={false}
    className="item"
    data-test-id="details"
    name="details"
    onClick={[Function]}
    role="option"
  >
    View Trade Details
  </div>
  <div
    aria-disabled={false}
    className="item"
    data-test-id="viewLc"
    name="viewLc"
    onClick={[Function]}
    role="option"
  >
    View LC Details
  </div>
</div>
`;

exports[`ExtraOptionsMenu as buyer ApplyLC shows 1`] = `
<div
  className="menu transition"
>
  <div
    aria-disabled={false}
    className="item"
    data-test-id="details"
    name="details"
    onClick={[Function]}
    role="option"
  >
    View Trade Details
  </div>
  <div
    aria-disabled={false}
    className="item"
    data-test-id="applyforLc"
    name="applyforLc"
    onClick={[Function]}
    role="option"
  >
    Apply for LC
  </div>
  <div
    aria-disabled={false}
    className="item"
    data-test-id="applyfornewSBLC"
    name="applyfornewSBLC"
    onClick={[Function]}
    role="option"
  >
    Apply for SBLC
  </div>
</div>
`;

exports[`ExtraOptionsMenu as buyer Reapply for LC is disabled 1`] = `
<div
  className="menu transition"
>
  <div
    aria-disabled={false}
    className="item"
    data-test-id="details"
    name="details"
    onClick={[Function]}
    role="option"
  >
    View Trade Details
  </div>
  <div
    aria-disabled={false}
    className="item"
    data-test-id="applyforLc"
    name="applyforLc"
    onClick={[Function]}
    role="option"
  >
    Apply for LC
  </div>
  <div
    aria-disabled={false}
    className="item"
    data-test-id="applyfornewSBLC"
    name="applyfornewSBLC"
    onClick={[Function]}
    role="option"
  >
    Apply for SBLC
  </div>
</div>
`;

exports[`ExtraOptionsMenu as buyer Reapply for LC shows if status is issued lc rejected 1`] = `
<div
  className="menu transition"
>
  <div
    aria-disabled={false}
    className="item"
    data-test-id="details"
    name="details"
    onClick={[Function]}
    role="option"
  >
    View Trade Details
  </div>
  <div
    aria-disabled={false}
    className="item"
    data-test-id="reapplyforLc"
    name="reapplyforLc"
    onClick={[Function]}
    role="option"
  >
    Reapply for LC
  </div>
  <div
    aria-disabled={false}
    className="item"
    data-test-id="applyfornewSBLC"
    name="applyfornewSBLC"
    onClick={[Function]}
    role="option"
  >
    Apply for SBLC
  </div>
</div>
`;

exports[`ExtraOptionsMenu as buyer Reapply for LC shows if status is request rejected 1`] = `
<div
  className="menu transition"
>
  <div
    aria-disabled={false}
    className="item"
    data-test-id="details"
    name="details"
    onClick={[Function]}
    role="option"
  >
    View Trade Details
  </div>
  <div
    aria-disabled={false}
    className="item"
    data-test-id="reapplyforLc"
    name="reapplyforLc"
    onClick={[Function]}
    role="option"
  >
    Reapply for LC
  </div>
  <div
    aria-disabled={false}
    className="item"
    data-test-id="applyfornewSBLC"
    name="applyfornewSBLC"
    onClick={[Function]}
    role="option"
  >
    Apply for SBLC
  </div>
</div>
`;

exports[`ExtraOptionsMenu as buyer View LC Details is disabled 1`] = `
<div
  className="menu transition"
>
  <div
    aria-disabled={false}
    className="item"
    data-test-id="details"
    name="details"
    onClick={[Function]}
    role="option"
  >
    View Trade Details
  </div>
  <div
    aria-disabled={false}
    className="item"
    data-test-id="applyforLc"
    name="applyforLc"
    onClick={[Function]}
    role="option"
  >
    Apply for LC
  </div>
  <div
    aria-disabled={false}
    className="item"
    data-test-id="applyfornewSBLC"
    name="applyfornewSBLC"
    onClick={[Function]}
    role="option"
  >
    Apply for SBLC
  </div>
</div>
`;

exports[`ExtraOptionsMenu as buyer View LC Details shows 1`] = `
<div
  className="menu transition"
>
  <div
    aria-disabled={false}
    className="item"
    data-test-id="details"
    name="details"
    onClick={[Function]}
    role="option"
  >
    View Trade Details
  </div>
  <div
    aria-disabled={false}
    className="item"
    data-test-id="viewLc"
    name="viewLc"
    onClick={[Function]}
    role="option"
  >
    View LC Details
  </div>
</div>
`;

exports[`ExtraOptionsMenu as buyer or seller shows view Trade 1`] = `
<div
  className="menu transition"
>
  <div
    aria-disabled={false}
    className="item"
    data-test-id="details"
    name="details"
    onClick={[Function]}
    role="option"
  >
    View Trade Details
  </div>
  <div
    aria-disabled={false}
    className="item"
    data-test-id="applyforLc"
    name="applyforLc"
    onClick={[Function]}
    role="option"
  >
    Apply for LC
  </div>
  <div
    aria-disabled={false}
    className="item"
    data-test-id="applyfornewSBLC"
    name="applyfornewSBLC"
    onClick={[Function]}
    role="option"
  >
    Apply for SBLC
  </div>
</div>
`;
